### ISSUE TITLE
Use an altimeter to keep constant altitude

### ISSUE THREAD

**Issue by lobodol** [lobodol/drone-flight-controller#7]
Use the BPM180 sensor to estimate altitude and keep it stable.

**Comment by officer47p** [comment]
> Use the BPM180 sensor to estimate altitude and keep it stable.

I don't think it has enough accuracy for that purpose. As I know it has a 0.25 meter of accuracy which isn't enough.
And as I guess, room pressure is different from the pressure outside of the room (house).
I think it's better to use an ultrasonic sensor.

**Comment by lobodol** [comment]
Yep, you're right, the accuracy seems to be not enough.
But I'm not sure an ultrasonic sensor would be better. If the drone flies over a house or a building, the measured distance from the floor will increase suddenly and the drone may leap. Do you agree ?

**Comment by officer47p** [comment]
> Yep, you're right, the accuracy of this sensor seems to be not enough.
> But I'm not sure an ultrasonic sensor would be better. If the drone flies over a house or a building, the measured distance from the floor will increase suddenly and the drone may leap. Do you agree ?

Yeah, that's true. I even have some problems with walls that are slightly tilted.
So we can make use of both of them!
Ultrasonic is our main reference, but when the drone faces an enormous change in the sensor, it checks if the pressure sensor reads the same pressure as before the leap, if it's the same, it subtracts the new distance from the old one. and then it can keep its altitude even when passing buildings.
We can use other sensors like Lidar sensor but it costs a lot.
Or we can have more ultrasonic sensors or pressure sensors and then take their average value.
GPS might help but hasn't enough accuracy even as BPM180.
I'm still not sure how DJI drones keep their height stable, I think they use some kind of image processing for that purpose.

**Comment by lobodol** [comment]
In my opinion, a sonar or any telemetric sensor would be great but only for lift-off and landing.
Imagine your drone has an altitude of 340m, it would take about 2s to the sonar to read the distance from the ground. This is too long. And telemeters having a such measure range would be very expensive.
I keep thinking an altimeter would be a good solution. Now, the question is: what is the needed accuracy ? In my mind, an accuracy until 5cm is acceptable.

**Comment by HungryNerd-exe** [comment]
Hi, Did you guys try implementing the altimeter using Barometer?

**Comment by lobodol** [comment]
Hi @HungryNerd-exe,  not yet

**Comment by 123sashank321** [comment]
> Use the BPM180 sensor to estimate altitude and keep it stable.
Hello @lobodol  first of all thank you very much for your amazing work .Can you please tell in detail how to use BPM 180 sensor with arduino Uno, required code and to which pins it should be connected..... (It will help me alot)  thank you.....
